### PR TITLE
v4.0.0

### PR DESCRIPTION
**Pull request by ppetrosh**
- Serial/then rework.

### CODE REVIEW

**Comment by ppetrosh** [comment]
@betzrhodes can you please review the changes and let me know if there are any examples missing in the main README. Thanks!

File: Promise.lib.nut [hunk 1]
```diff
@@ -52,49 +51,49 @@ class Promise {
...
      */
     function _callHandlers() {
-        if (this.STATE_PENDING != this._state) {
+        if (STATE_PENDING != _state) {
             imp.wakeup(0, function() {
-                if (this._last && this._handlers.len() == 0
-                               && this.STATE_REJECTED == this._state) {
-                    server.log(PROMISE_ERR_UNHANDLED_REJ + this._value);
+                if (_last && _handlers.len() == 0
+                               && STATE_REJECTED == _state) {
```
Comment (betzrhodes) [inline Promise.lib.nut]: formatting - indenting is off.

Comment (ppetrosh) [reply Promise.lib.nut]: Yeah, fixed. Wasn't mine though :)

**Comment by betzrhodes** [comment]
Most code comments are multi-line. Should these be updated?

File: README.md [hunk 2]
```diff
@@ -129,40 +169,88 @@ Promise.reject(reason)
...
 ### Promise.race(*series*)
 
-This method executes promises in parallel and resolves when the first is done. It returns a promise that resolves or rejects with the first resolved/rejected promise value.
+This method executes promises in parallel and resolves when the first is done. 
+It returns a promise that resolves or rejects with the first resolved/rejected promise value.
 
 The parameter *series* is an array of promises and/or functions that return promises.
 
-For example, in the following code *p* rejects with value `"1"` in 1 second:
+For example, in the following code *p* rejects with value `"1"` in 0.5 second:
```
Comment (betzrhodes) [inline README.md]: this is not correct - it rejects with value "3" in 0.5s

**Comment by ppetrosh** [comment]
> Most code comments are multi-line. Should these be updated?

Sure, done. Thanks.

File: README.md (hunk 2 above)
Comment (ppetrosh) [reply README.md]: Oh, value 3, yes. Fixed, thanks!

Comment (betzrhodes) [reply README.md]: yes - condition 3 rejects first not 1.

Comment (betzrhodes) [reply README.md]: condition 3 is 0.5s condition 1 is 1s

Comment (betzrhodes) [reply README.md]: also the comment in the code should be updated.

File: README.md [hunk 3]
```diff
@@ -205,45 +293,87 @@ This method returns a promise that resolves when all the promises in the chain r
 
 The parameter *series* is an array of promises and/or functions that return promises.
 
-For example, in the following code *p* rejects with value `"2"` in 2.5 seconds:
+The action function is triggered at the moment when the Promise instance is created. So using functions returning Promise instances to pass into `Promise.serial` makes instantiation sequential. I.e. a promise is created and the action is triggered only when the previous Promise in the series got resolved or rejected.
+
+For example, in the following code `p` rejects with value `"2"` in 2.5 seconds
```
Comment (betzrhodes) [inline README.md]: As written, the series doesn't reject - all promises resolve. The the value that p resolves with is "3" in 2.5 seconds.  

I added logs and measured time using harware.millis. Here are the logs I get when I run the example.

2018-08-17 13:09:30 -07:00 | [Device] | starting...
-- | -- | --
2018-08-17 13:09:30 -07:00 | [Device] | p1 - scheduling 1s timer
2018-08-17 13:09:30 -07:00 | [Device] | p3 - scheduling 0.5s timer
2018-08-17 13:09:31 -07:00 | [Device] | p3 0.5s timer expired
2018-08-17 13:09:31 -07:00 | [Device] | 0.508s from start
2018-08-17 13:09:31 -07:00 | [Device] | p1 1s timer expired
2018-08-17 13:09:31 -07:00 | [Device] | 1.012s from start
2018-08-17 13:09:31 -07:00 | [Device] | p2 - scheduling 1.5s timer
2018-08-17 13:09:33 -07:00 | [Device] | p2 1.5s timer expired
2018-08-17 13:09:33 -07:00 | [Device] | 2.528s from start
2018-08-17 13:09:33 -07:00 | [Device] | Resolved with value: 3
2018-08-17 13:09:33 -07:00 | [Device] | 2.535s from start

File: README.md [hunk 4]
```diff
@@ -205,45 +293,87 @@ This method returns a promise that resolves when all the promises in the chain r
...
-### TL;DR
-
-```bash
-cp .imptest .imptest-local
-nano .imptest-local # edit device/model
-imptest test -c .imptest-local
-```
-
-### Running Tests
+While in the following code p rejects in 1.5 seconds as all the Promises are instantiated at the same time:
```
Comment (betzrhodes) [inline README.md]: again this example doesn't reject, it resolves in 1.5 seconds with a value of 3.

**Comment by betzrhodes** [comment]
We still don't have good examples/documentation around syntax for async promises that wait before firing the next action. The updated series example is a start, but it is still a pretty confusing example. We are still missing documentation on how to make .then functions with resolve/reject calls inside an imp.wakeup call wait for the timer to expire before before continuing the next action.  

This factory code has a good example of async promises that wait for completion before continuing. Can probably simplify and add to examples folder. 
https://beta-impcentral.electricimp.com/ide/products/e314c893-e94f-7857-360c-8aa8406d25a1/devicegroups/d307b125-3576-e0b8-0eac-789d5b2287ff

File: README.md (hunk 3 above)
Comment (ppetrosh) [reply README.md]: yes, fixed. thanks!

File: README.md (hunk 4 above)
Comment (ppetrosh) [reply README.md]: you are right. fixed.

**Comment by ppetrosh** [comment]
> We still don't have good examples/documentation around syntax for async promises that wait before firing the next action. The updated series example is a start, but it is still a pretty confusing example. 

I've spent a lot of time on this but see no way to make it easier (other than by using lambdas). How would you simplify these examples and make then less confusing?

> We are still missing documentation on how to make .then functions with resolve/reject calls inside an imp.wakeup call wait for the timer to expire before continuing the next action.

Then function doesn't have resolve/reject parameters. It's all synchronous. There are no good ways to implement a series of synchronous then calls with asynchronous actions triggered by each of them. It just doesn't work this way.

I really spent some time on how to make it clear, but don't see how to improve the docs. The rule is simple: 

|  Actions  | Execution   | Suggestions  | Examples |
|---|---|---|---|
| Sync   | Sync  | `.then`,<br>`.serial`, or<br> `.loop`  | |
| Async  | Sync   | `.serial` + `f() {return Promise(action)}`, <br> `.loop` + `f() {return Promise(action)}` or <br>`.then` + `return Promise(action)` | |
|  Async | Async  | `.all`,<br>`.loop`,<br>`.serial`, or<br>`.then` + `return Promise(action)`| |
|  Sync | Async  | N/A  | N/A |
|  Sync and Async | Sync | `.serial` + `f() {return Promise}` or <br>`.loop` + `f() {return Promise}` | |
|  Sync and Async | Async | N/A | N/A |

`Sync` - the series of Promises terminates if one of them rejects
`Async` - the series is done when all the promises are done (regardless of whether they are resolved or rejected).

This qualification has a number of limitations:

1. it doesn't distinguish cases when you need results from ALL the promises or just from the last one
2. it doesn't address cases when Async needs to resolve when any of the promises rejects (first one)

So it's way much more complicated than that.

Having said that, I would probably add this table to the README. But there is no simple set of simple use cases that we can provide as I can see it.

**Comment by betzrhodes** [comment]
Having many of the very simple examples with logs, showing the timing of functions triggering would be helpful. 
Having some instructions about the difference between returning vs just calling resolve/reject would be a good start.
I think right now we don't have enough examples showing usage. Example c is super useful - that is what I used to figure out how to write factory test code. But this example really only shows one way to use promises. There are not good full examples showing uses for .all and .serial - having fuller examples showing actual use case for each of these would be nice.

**Comment by ppetrosh** [comment]
Please read the updated comment above https://github.com/electricimp/Promise/pull/33#issuecomment-413988612

**Comment by betzrhodes** [comment]
I agree it is complicated, I just don't think we have a good set of examples yet. Every time I use promises I spend a lot of time playing around until I get them to work the way I need. It would be nice if our docs or examples helped more. For one - If a promise is returned it behaves very differently then just creating a promise - info on this (with a couple contrasting examples) would be very helpful.

**Comment by ppetrosh** [comment]
I just realized the table is not complete... Let me finish it and I agree, it probably makes sense to add examples for each of the rows.

**Comment by ppetrosh** [comment]
Check https://github.com/electricimp/Promise/pull/33#issuecomment-413988612 out please. I think I'll add separate examples (with links from the table) for each of the rows in the "Suggestions" (should be renamed) column. Does it sound complete to you now?

**Comment by betzrhodes** [comment]
Yes this table helps a ton! I think this is a big improvement.

**Comment by ppetrosh** [comment]
Ok, will take it from here. Thanks for the discussion. It helped me to make it clear for myself.

**Pull request merged**